sql: prohibit ALTER TABLE SET DATA TYPE requiring conversion/validation

otan opened the pull request:
Refs: #54844 

sql: prohibit ALTER TABLE SET DATA TYPE requiring conversion/validation

Performing the validation and conversion as we do today in ALTER TABLE
... SET DATA TYPE ... is broken as it does not perform certain precision
or width based checks correctly.

The fix for this is deemed risky, hence we are disabling the
functionality in v20.2. This commit is intended for backport only.

Release note (sql change): ALTER TABLE ... SET DATA TYPE ... is no
longer available for operations that involve conversion (e.g. string ->
timestamptz) or precision/width truncation (e.g. INT(4) -> INT(2)) due
to a bug in validation. These features are already gated by the session
variable `enable_experimental_alter_column_type_general` - setting this
session variable is now a no-op.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/56621)
<!-- Reviewable:end -->

otan commented:
cc @solongordon i'm intending to backport the approved PR onto 20.2, and close this one if it looks ok

otan committed:
sql: prohibit ALTER TABLE SET DATA TYPE requiring conversion/validation

Performing the validation and conversion as we do today in ALTER TABLE
... SET DATA TYPE ... is broken as it does not perform certain precision
or width based checks correctly.

The fix for this is deemed risky, hence we are disabling the
functionality in v20.2. This commit is intended for backport only.

Release note (sql change): ALTER TABLE ... SET DATA TYPE ... is no
longer available for operations that involve conversion (e.g. string ->
timestamptz) or precision/width truncation (e.g. INT(4) -> INT(2)) due
to a bug in validation. These features are already gated by the session
variable `enable_experimental_alter_column_type_general` - setting this
session variable is now a no-op.
